File: FloodFill.java
class Solution {
    public int[][] floodFill(int[][] image, int sr, int sc, int newColor) {
        // Code here
        if(image[sr][sc] == newColor){
            return image;
        }
        int n = image.length;
        int m = image[0].length;
        floodFillUtils(image, image[sr][sc], sr, sc, n, m, newColor);
        return image;
    }
    
    void floodFillUtils(int[][] image, int origColor, int sr, int sc, int n, int m, int newColor){
        if(sr < 0 || sc < 0 || sr >= n || sc >= m){
            return;
        }
        
        
        if(image[sr][sc] == newColor || image[sr][sc] != origColor){
            return;
        }
        
        image[sr][sc] = newColor;
        floodFillUtils(image, origColor, sr+ 1, sc, n, m, newColor);
        floodFillUtils(image, origColor, sr - 1, sc, n, m, newColor);
        floodFillUtils(image, origColor, sr, sc - 1, n, m, newColor);
        floodFillUtils(image, origColor, sr, sc + 1, n, m, newColor);
    } 
}

File: FloodFill.cpp
class Solution {
  public:
    
    vector<vector<int>> floodFill(vector<vector<int>>& image, int sr, int sc, int newColor) {
        if(image[sr][sc] == newColor)
            return image;

        int n = image.size();
        int m = image[0].size();

        dfs(image, image[sr][sc], newColor, sr, sc, n, m);
        return image;
    }

    void dfs(vector<vector<int>>& image, int origColor, int newColor, int i, int j, int n, int m) {
        if(i < 0 || j < 0 || i >= n || j >= m)
            return;

        if(image[i][j] == newColor || image[i][j] != origColor)
            return;

        image[i][j] = newColor;
        dfs(image, origColor, newColor, i + 1, j, n, m);
        dfs(image, origColor, newColor, i - 1, j, n, m);
        dfs(image, origColor, newColor, i, j - 1, n, m);
        dfs(image, origColor, newColor, i, j + 1, n, m);
    }

};
